tests/integration/retries.py

```diff
@@ -0,0 +1,214 @@
+                )
+            elif (timeout * 0.80) <= attempt_duration:
+                logger.warning(
+                    f"Retryable call ({func.__name__}) attempt took {attempt_duration}, within 20% of the specified timeout ({timeout}); please adjust timeout accordingly."
+                )
+
+    def _check_retries_exhausted(self, attempt: int, deadline: float | None, pending_sleep: float) -> str | None:
+        if self._min_attempts < attempt:
+            max_attempts = self._max_attempts
+            if max_attempts is not None and max_attempts < attempt:
+                return f"Timed out after exceeding maximum number of attempts ({self._max_attempts})"
+            if deadline is not None:
+                now = self._clock.time()
+                if deadline < now:
+                    return f"Timed out after {self._timeout}"
+                if (deadline - pending_sleep) < now:
+                    return f"Will time out after {self._timeout}"
+        return None
+
+    def decorator(self, func: Callable[P, T]) -> Callable[P, T]:
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs) -> T:
+            deadline = self._deadline()
+            attempt = 1
+            while True:
+                started_attempt_at = self._clock.time()
+                try:
+                    result = func(*args, **kwargs)
+                    completed_attempt_at = self._clock.time()
+                    attempt_duration = dt.timedelta(seconds=completed_attempt_at - started_attempt_at)
+                    self._check_attempt_duration(func, attempt_duration)
+                    return result
+                except Exception as err:  # pylint: disable=broad-exception-caught
+                    last_err = err
+                    # sleep 10s max per attempt, unless it's HTTP 429 or 503
+                    retry_after_secs = getattr(err, 'retry_after_secs', None)
+                    if retry_after_secs is not None:
+                        # cannot depend on DatabricksError directly because of circular dependency
+                        retry_reason = 'throttled by platform'
+                        sleep = retry_after_secs
+                    else:
+                        maybe_retry_reason = self._can_retry(err)
+
+                        if maybe_retry_reason is None:
+                            # raise if exception is not retryable
+                            raise err
+                        retry_reason = maybe_retry_reason
+                        sleep = min(10, attempt)
+
+                attempt += 1
+                logger.debug(f'Retrying (attempt={attempt}): {retry_reason} (sleeping ~{sleep}s)')
+
+                # Save some time upon failure by failing early if the deadline will be exceeded after sleeping.
+                sleep_duration = sleep + random()
```

Comment:
Maybe move this into the `_check_retries_exhausted`

Reply:
Btw this check if the timeout will happen in less than a second right? How likely is that? I recon very unlikely

Reply:
The sleep itself can be up to 11 seconds, but can be higher if the attempt is due to (upstream) throttling by the platform.

This check is about not bothering to do the sleep if we know we're to fail immediately afterwards. The alternative here, and what the original code did, is to do the `sleep()` call here and then check afterwards if the we're past the deadline.

Reply:
But does the `random()` increase the likelihood of early catching a timeout significantly?

Reply:
In the literal sense: given a timeout _could_ be caught early it increases the likelihood of it being caught by between 9% and 50%.

That said I think the more important thing to consider here is whether it's worth trying to catch it early at all because it would be simpler to simply sleep and then check the timeout afterwards.